### PR TITLE
Recursive tasklist added

### PR DESCRIPTION
**Pull request by WloBeb**
Summary, after all commits:
- new recursive tasklist concept implemented
- new flag "recursiveList" to mark new type tasklist (only these lists  are "expanded")
- new flag "useMassTask" - profiles with this flag "on" use "_Mass" task when possible
- new tasklists "mass refining" for every profession (use of <b>mass gathering</b>, when the components are missing)

### CODE REVIEW

**Comment by noonereally** [comment]
Please read about the addToDefined function, i think it should be changed a bit:
https://github.com/noonereally/NW-Profession-Bot/issues/19

(not sure how the notification works with name tags)

**Comment by WloBeb** [comment]
@BigRedBrent 
No, addProfileToDefined() function don't support recursive list now.
but you can write after "default" profie:

<pre>          25 : ["Crafted_Med_Armorsmithing_T4_Refine_Basic", "Crafted_Med_Armorsmithing_T4_Gather_Basic"],
            }
        }<i><b>,{
            profileName: "Wondrous Sprocket",
            isProfileActive: false,
            recursiveList : true,
            level: {
                0 : ["Med_Armorsmithing_Tier0_Intro"],
                1 : ["Med_Armorsmithing_Tier1_Gather_Basic"],
                6 : ["Med_Armorsmithing_Tier1_Event_Gond"],
            },
        }</b></i>]</pre>

If you don't notice that, after this definition:

<pre>        {
            profileName : "mass refining",
            isProfileActive : true,
            recursiveList : true,
            level : {
                0 : ["Alchemy_Tier0_Intro_1"],
                1 : ["Alchemy_Tier1_Refine_Basic_Mass"],
                7 : ["Alchemy_Tier2_Refine_Basic_Mass"],
                14 : ["Alchemy_Tier3_Refine_Basic_Mass"],
                21 : ["Alchemy_Tier4_Refine_Basic_Mass"],
        },</pre>

tasklist for level 25: ["Alchemy_Tier4_Refine_Basic_Mass", "Alchemy_Tier3_Refine_Basic_Mass", "Alchemy_Tier2_Refine_Basic_Mass","Alchemy_Tier1_Refine_Basic_Mass"]

It is <b>concatenation</b> of all lower level task (from lvl 1), not duplication.

**Comment by dlebedynskyi** [comment]
https://github.com/noonereally/NW-Profession-Bot/issues/19
@noonereally @WloBeb @BigRedBrent 
Here is a fiddle with recursive list add. 

https://jsfiddle.net/63f7mpf6/10/

also simplified new profession add.
Ill add it shortly after testing. If you find something - report pls

**Comment by noonereally** [comment]
About the "EDIT: Removed exclusion of mass tasks for profileNames begining with "mass "
So, we can use mass refining tasks in profile, and mass gathering should be selected, if we will be missing ingredients."

I think a profile flag to exclude mass or include will be better than counting on names, like the recursion flag.

**Comment by dlebedynskyi** [comment]
Pull is ready to merge. Lets not mix everything in one pile. Clean up Mass later.

**Comment by WloBeb** [comment]
@noonereally say: "I think a profile flag to exclude mass or include will be better than counting on names, like the recursion flag."

You're probably right but profile isn't easy accessible in this place of code. 

EDIT: Done :D.

File: main.js [hunk 1]
```diff
@@ -1865,6 +1961,35 @@ function _select_Gateway() { // Check for Gateway used to
         },
     });
 
+    // expand recursive tasklist
+    for (var professionIdx in definedTask) {
```
Comment (Phr33d0m) [inline main.js:1965]: This n^4+xn nightmare scares me, and it should scare you too.

Comment (noonereally) [inline main.js:1965]: You have at most 13 professions ?

Comment (WloBeb) [inline main.js:1965]: Shouldn't scare me, if I know how large this array is :D
Are you scared to: 

<pre>for (var i=0; i<13; i++)
  for (var j=0; j<20; j++)</pre>

Comment (WloBeb) [inline main.js:1965]: Measured with current tasklists: 82.131 ms :D on very slow netbook.
And this is done only once after start.

Edit:
Next run: this loop 43.585 ms. Compared with addSettings: 10658.668ms and SaveSettings 5727,825ms, definitely doesn't scare me.

BTW. Settings is formally 1-dimension array, but contains at least "number of profession" \* "number of character" \* "number of fields per character" fields :P  Is it not scare you?

Comment (WloBeb) [inline main.js:1965]: For me complexity of these loops is n^2 (unless I don't know something about the complexity of javascript 'in' implementation).
And that for n equal to approximately 10-20 (in worst case).

Compared to compexity n^3 with time-consuming GM_ operations.

I understood, no more "Pull requests". Why should I waste my and your time...

Comment (noonereally) [inline main.js:1965]: If to be precise it's O(n) where n is the number of total levels in all the profiles. 
Or O(nm) where m is the number of items in a level.
If to be practical it's maximum 13 * 10 * 25 * 5, and sequential reading.
This is not important.

ps. `profile.level[i].indexOf(profile.level[i-1][j])` is a loop too.
But as i said, imo, totally insignificant. A single pass on all of it wont change anything.
The GM_save is an issue though.

**Comment by WloBeb** [comment]
@noonereally As I said: "I understood, no more "Pull requests". Why should I waste my and your time..."
My changes are in my branch...
BTW. Did you realised that @Phr33d0m authored commits from @dlebedynskyi, which probably doesn't work correctly without changes from my commits? And for me he has only stupid O(n^4) comment :/

**Comment by WloBeb** [comment]
Do with it what you want..

**Comment by Phr33d0m** [comment]
@WloBeb I was testing @dlebedynskyi's changes meanwhile. Nobody has told you "no more pull-reqs".

What's more, I initially thought you both made different implementations of the same idea and as @dlebedynskyi's one seemed cleaner and easier to understand I decided to pick it up and test it out, which as far is working pretty well.

Can you elaborate more on what isn't working correctly from his implementation? So I can apply your changes accordingly.

**Comment by dlebedynskyi** [comment]
@WloBeb  how old are you? cause you behave like a 10 year old when somebody does not appreciate this drawing. If you have issues, tell what is wrong and what can be done, otherwise get over it

**Comment by Phr33d0m** [comment]
@dlebedynskyi please abstain from such comments. As an adult you should already know that your comment would only irritate people and start unnecessary drama. Something I won't approve here.

You both implemented different ways of the same idea and which one is better and why should be discussed over politely and respectfully. I picked one up only by sight-reading the code, I might be wrong and a different approach might be neccessary. That's something we can see only through testing.

**Comment by WloBeb** [comment]
@Phr33d0m These changes are obsolete.
I rewrote all stuff related to task definition:
- new addProfile() function which can add 'clean' profile or override existing one (it includes expanding of tasklist but could be easy changed to @noonereally idea if it removes)
  EDIT: expansion may concern only non-empty levels :D
- in code first definedTask definitions then added profiles.

Everything in my 'master' branch.

When i saw O(n^4) felt like a boy on your profile picture.

**Comment by Phr33d0m** [comment]
@WloBeb thanks. I'll check it out. On first sight I really really like how all the `definedTask[]` and `addProfile()` are grouped. Also, I like your `addProfile()` much more now, even if there are quite a bit of stuff that can be polished.

Also, have a look at what's going on here: noonereally#17 (and this commit: https://github.com/dlebedynskyi/NW-Profession-Bot/commit/36362e507b4258c38e3d462b66ddac7cc0ba9e60). The 3 of you are trying to implement the same thing, so in the end one (or two) of you would be doing unnecessary coding so try to figure it out together and work as a group :-).

**Comment by WloBeb** [comment]
@Phr33d0m In <a href=https://github.com/WloBeb/NW-Profession-Bot/commit/5463c0444934e42008b8b419a6a61c49a6258024>this</a> commit I probably joined my & @noonereally ideas :)
BTW. its <b>my</b> first addProfile() implementation. Previous were from @dlebedynskyi

**Comment by WloBeb** [comment]
Close this pull request because this code is partialy obsolete - see  comment  above.

**Pull request closed without merging**